src/messages/layouts/MessageLayoutContainer.cpp

```diff
@@ -456,6 +457,45 @@ size_t MessageLayoutContainer::getFirstMessageCharacterIndex() const
     return index;
 }
 
+std::pair<int, int> MessageLayoutContainer::getWordBounds(
+    const MessageLayoutElement *hoveredElement) const
+{
+    size_t index = 0;
+    size_t wordStart = 0;
+
+    for (; index < this->elements_.size(); index++)
+    {
+        const auto &element = this->elements_[index];
+        if (element->getWordId() == hoveredElement->getWordId())
+        {
+            break;
+        }
+
+        wordStart += element->getSelectionIndexCount();
+    }
+
+    size_t wordEnd = wordStart;
+
+    for (; index < this->elements_.size(); index++)
+    {
+        const auto &element = this->elements_[index];
+        if (element->getWordId() != hoveredElement->getWordId())
+        {
+            break;
+        }
+
+        wordEnd += element->getSelectionIndexCount();
+    }
+
+    const auto *lastElementInSelection = this->elements_[index - 1].get();
```

Comment:
Not sure how feasible it is, but could you add a unit test that runs getWordBounds on a MessageLayoutContainer with some edge cases? like 0 elements, 1 element, hoveredElement doesn't exist

Specifically the `index - 1` part to me is scary

Reply:
Don't know if there is a way to create an empty `MessageLayoutContainer`, but I like having a failsafe here. I will add it and see what I can do about the tests

Reply:
I battled my way through gtest and somehow managed to make a working test. I left some comments which I think you should take a look at. I'll add more tests and fix some of the stuff pointed out tomorrow